### PR TITLE
Fixed address Support for bloxone Ansible V2

### PR DESCRIPTION
**Pull request by JchhatbarInfoblox**
Added Fixed address Support for bloxone Ansible V2

### CODE REVIEW

File: changelogs/fragments/38-fixed-address.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+deprecated_features:
+  - b1_ipam_fixed_address - is deprecated in favor of `ipam_fixed_address`.
+  - b1_ipam_fixed_address_gather - is deprecated in favor of `ipam_fixed_address_info`.
```
Comment (unasra) [inline changelogs/fragments/38-fixed-address.yml:3]: Please add a line at the end of the file .

File: plugins/modules/b1_ipam_fixed_address.py [hunk 2]
```diff
@@ -432,6 +436,11 @@ def main():
     }
 
     module = AnsibleModule(argument_spec=argument_spec)
+    module.deprecate(
+        "This module is being deprecated and will be removed in version 3.0.0. Use the M(ipam_fixed_address) module instead.",
+        version="3.0.0",
+        collection_name="infoblox.bloxone",
+    )
```
Comment (mathewab) [inline plugins/modules/b1_ipam_fixed_address.py:443]: Why was this needed? 

I am asking, since this is not mentioned in https://docs.ansible.com/ansible/latest/dev_guide/module_lifecycle.html#deprecating-modules
Also, we haven't done this for the other deprecated modules. Will like to be consistent if possible.

File: tests/integration/targets/ipam_fixed_address/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        name: "{{ name }}"
+        state: "present"
+      register: ip_space
+
+    - name: "Create a Subnet"
+      infoblox.bloxone.ipam_subnet:
+        address: "10.0.0.0/24"
+        space: "{{ ip_space.id }}"
+        state: "present"
+      register: subnet
```
Comment (mathewab) [inline tests/integration/targets/ipam_fixed_address/tasks/main.yml:25]: It might be better to separate this out as `setup_ip_space` and `setup_subnet` and add as dependencies. This will allow us to reuse these targets.

File: tests/integration/targets/ipam_fixed_address/tasks/main.yml
Comment (mathewab) [inline tests/integration/targets/ipam_fixed_address/tasks/main.yml:1]: We need more tests here, ideally one per "writable" sub field. If a particular sub field can't be tested, it is also better to list them out at the beginning of the file as a TODO

File: plugins/modules/ipam_fixed_address_info.py [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+        description:
+            - Filter query to filter objects by tags
+        type: str
+        required: false
+
+extends_documentation_fragment:
+    - infoblox.bloxone.common
+"""  # noqa: E501
+
+EXAMPLES = r"""
```
Comment (unasra) [inline plugins/modules/ipam_fixed_address_info.py:59]: Example are missing here

File: plugins/modules/ipam_fixed_address.py [hunk 5]
```diff
@@ -0,0 +1,783 @@
...
+        type: "option"
+    disable_dhcp: false
+    header_option_filename: "filename"
+    header_option_server_address: "server_address"
+    header_option_server_name: "server_name"
+    hostname: "hostname.example.com"
+    inheritance_parent: "parent_id"
+    inheritance_sources:
+      dhcp_options:
+        action: "inherit"
```
Comment (unasra) [inline plugins/modules/ipam_fixed_address.py:223]: Might be an indentation error here .
Is the example used here tested ??

File: plugins/modules/ipam_fixed_address.py [hunk 6]
```diff
@@ -0,0 +1,783 @@
...
+EXAMPLES = r"""
+- name: Create a fixed address
+  infoblox.bloxone.ipam_fixed_address:
+    address: "10.0.0.1"
+    ip_space: "example_ip_space"
+    match_type: "mac"
+    match_value: "00:00:00:00:00:00"
+    state: "present"
+
+- name: Create a fixed address with all parameters
```
Comment (unasra) [inline plugins/modules/ipam_fixed_address.py:204]: Have an Example with all params isn't necessary. The idea is to basically add all the fields that generally used

File: plugins/modules/ipam_fixed_address.py [hunk 7]
```diff
@@ -0,0 +1,783 @@
...
+        option_code: "code1"
+        option_value: "value1"
+        type: "option"
+    disable_dhcp: false
+    header_option_filename: "filename"
+    header_option_server_address: "server_address"
+    header_option_server_name: "server_name"
+    hostname: "hostname.example.com"
+    inheritance_parent: "parent_id"
+    inheritance_sources:
```
Comment (unasra) [inline plugins/modules/ipam_fixed_address.py:221]: The API currently requires all fields inside the inheritance config to be explicitly provided .

File: plugins/modules/ipam_fixed_address.py [hunk 8]
```diff
@@ -0,0 +1,783 @@
...
+        action: "inherit"
+      header_option_server_address:
+        action: "inherit"
+      header_option_server_name:
+        action: "inherit"
+    ip_space: "example_ip_space"
+    match_type: "mac"
+    match_value: "00:00:00:00:00:00"
+    name: "fixed_address_name"
+    parent: "parent_id"
```
Comment (unasra) [inline plugins/modules/ipam_fixed_address.py:236]: Parent should be a read only field

File: plugins/modules/ipam_fixed_address.py [hunk 9]
```diff
@@ -0,0 +1,783 @@
...
+  infoblox.bloxone.ipam_fixed_address:
+    address: "10.0.0.1"
+    ip_space: "example_ip_space"
+    match_type: "mac"
+    match_value: "00:00:00:00:00:00"
+    state: "present"
+
+- name: Create a fixed address with all parameters
+  infoblox.bloxone.ipam_fixed_address:
+    id: "fixed_address_id"
```
Comment (unasra) [inline plugins/modules/ipam_fixed_address.py:206]: `id` is a read only parameter

File: plugins/modules/ipam_fixed_address.py [hunk 10]
```diff
@@ -0,0 +1,783 @@
...
+    match_value: "00:00:00:00:00:00"
+    name: "fixed_address_name"
+    parent: "parent_id"
+    tags:
+      key1: "value1"
+      key2: "value2"
+
+- name: Delete a fixed address
+  infoblox.bloxone.ipam_fixed_address:
+    id: "fixed_address_id"
```
Comment (unasra) [inline plugins/modules/ipam_fixed_address.py:243]: `id` is a read only param. FA for Delete call should be identified using Primary Keys , namely the address and the IP Space

File: tests/integration/targets/ipam_fixed_address/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+    - name: "Create a Fixed Address"
+      infoblox.bloxone.ipam_fixed_address:
+        address: "10.0.0.1"
+        name: "test_fixed_address_ansible"
+        match_type: "mac"
+        match_value: "00:00:00:00:00:00"
+        ip_space: "{{ ip_space.id }}"
+        state: "present"
+      register: fixed_address
+
```
Comment (unasra) [inline tests/integration/targets/ipam_fixed_address/tasks/main.yml:36]: As per trend for previous objects , we add the tasks for creation/updation , get and assert under one block of code without any empty lines in between . This way empty lines can be used to indicate the next create/update ,read and assert code block

File: tests/integration/targets/ipam_fixed_address/tasks/main.yml [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+      csp_url: "{{ csp_url }}"
+      api_key: "{{ api_key }}"
+
+  block:
+    # Create a random IP space name to avoid conflicts
+    - ansible.builtin.set_fact:
+        name: "test-ip-space-{{ 999999 | random | string }}"
+        tag_value: "site-{{ 999999 | random | string }}"
+
+    # Basic tests for Subnet
```
Comment (unasra) [inline tests/integration/targets/ipam_fixed_address/tasks/main.yml:13]: Tests for "check mode" are missing.

File: tests/integration/targets/ipam_fixed_address/tasks/main.yml [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+        match_value: "00:00:00:00:00:00"
+        ip_space: "{{ ip_space.id }}"
+        state: "present"
+      register: fixed_address
+    - assert:
+        that:
+          - fixed_address is not changed
+          - fixed_address is not failed
+
+    - name: "Delete a Fixed Address"
```
Comment (unasra) [inline tests/integration/targets/ipam_fixed_address/tasks/main.yml:64]: To maintain consistency , we have a create , update delete cycle in tests where after the "check_mode" and "idempotent" tests , we create the object and update it with additional params and add a delete call at the end.

File: tests/integration/targets/ipam_fixed_address_info/tasks/main.yml [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+        state: "present"
+      register: ip_space
+
+    - name: "Create a Subnet"
+      infoblox.bloxone.ipam_subnet:
+        address: "10.0.0.0/24"
+        space: "{{ ip_space.id }}"
+        state: "present"
+      register: subnet
+
```
Comment (unasra) [inline tests/integration/targets/ipam_fixed_address_info/tasks/main.yml:26]: Use `setup_ip_space` and `setup_subnet` instead.

File: meta/runtime.yml [hunk 15]
```diff
@@ -18,6 +18,8 @@ action_groups:
     - ipam_subnet_info
     - ipam_address_block
     - ipam_address_block_info
+    - ipam_fixed_address
```
Comment (unasra) [inline meta/runtime.yml:21]: As per the swagger , its `dhcp_fixed_address`  and `dhcp_fixed_address_info` . this change should be made throughout the PR.

**Comment by JchhatbarInfoblox** [comment]
Closing PR as fixed address is part of DHCP.
